### PR TITLE
Web exceptions refactoring

### PR DESCRIPTION
**Pull request by asvetlov**
Don't inherit `web.HTTPException` from `Response`

### CODE REVIEW

File: aiohttp/web_exceptions.py [hunk 1]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
-                          headers=headers, reason=reason,
-                          body=body, text=text, content_type=content_type)
-        Exception.__init__(self, self.reason)
-        if self.body is None and not self.empty_body:
-            self.text = "{}: {}".format(self.status, self.reason)
+        if reason is None:
+            try:
+                reason = RESPONSES[self.status_code][0]
+            except Exception:
+                reason = ''
```
Comment (kxepal) [inline aiohttp/web_exceptions.py]: It feels like not a good reason to swallow the exception. However, I don't feel the intentions for this. Are they good?

File: aiohttp/web_exceptions.py [hunk 2]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
+
+        if text is None:
+            if not self.empty_body:
+                text = "{}: {}".format(self.status_code, reason)
+        else:
+            if self.empty_body:
+                warnings.warn(
```
Comment (kxepal) [inline aiohttp/web_exceptions.py:100]: For which cases this warning is corrent?

File: aiohttp/web_exceptions.py [hunk 3]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
-        Response.__init__(self, status=self.status_code,
-                          headers=headers, reason=reason,
-                          body=body, text=text, content_type=content_type)
-        Exception.__init__(self, self.reason)
-        if self.body is None and not self.empty_body:
-            self.text = "{}: {}".format(self.status, self.reason)
+        if reason is None:
+            try:
+                reason = RESPONSES[self.status_code][0]
+            except Exception:
```
Comment (webknjaz) [inline aiohttp/web_exceptions.py]: KeyError + IndexError?

File: aiohttp/web_exceptions.py [hunk 4]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
+                warnings.warn("content_type without text is deprecated",
+                              DeprecationWarning,
+                              stacklevel=2)
+            real_headers[hdrs.CONTENT_TYPE] = content_type
+        elif hdrs.CONTENT_TYPE not in real_headers and text:
+            real_headers[hdrs.CONTENT_TYPE] = 'text/plain'
+
+        super().__init__(reason)
+
+        self._reason = reason
```
Comment (kxepal) [inline aiohttp/web_exceptions.py]: This assignment looks strange. Shouldn't it be handled in super-classes?

File: aiohttp/web_exceptions.py [hunk 5]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
-                DeprecationWarning)
-        Response.__init__(self, status=self.status_code,
-                          headers=headers, reason=reason,
-                          body=body, text=text, content_type=content_type)
-        Exception.__init__(self, self.reason)
-        if self.body is None and not self.empty_body:
-            self.text = "{}: {}".format(self.status, self.reason)
+        if reason is None:
+            try:
+                reason = RESPONSES[self.status_code][0]
```
Comment (webknjaz) [inline aiohttp/web_exceptions.py]: Maybe add method for this with lru cache?

File: aiohttp/web_exceptions.py [hunk 6]
```diff
@@ -83,23 +88,71 @@ class HTTPException(Response, Exception):
...
+            except Exception:
+                reason = ''
+
+        if text is None:
+            if not self.empty_body:
+                text = "{}: {}".format(self.status_code, reason)
+        else:
+            if self.empty_body:
+                warnings.warn(
+                    ("text argument is deprecated for HTTP status {} ,"
```
Comment (webknjaz) [inline aiohttp/web_exceptions.py]: No need to have braces around text here

Comment (kxepal) [reply aiohttp/web_exceptions.py]: Don't we have yet black to handle such styling issues?

File: aiohttp/web_exceptions.py [hunk 7]
```diff
@@ -241,15 +297,22 @@ def __init__(self,
...
                  text: Optional[str]=None,
                  content_type: Optional[str]=None) -> None:
         allow = ','.join(sorted(allowed_methods))
         super().__init__(headers=headers, reason=reason,
-                         body=body, text=text, content_type=content_type)
+                         text=text, content_type=content_type)
         self.headers['Allow'] = allow
-        self.allowed_methods = set(allowed_methods)  # type: Set[str]
-        self.method = method.upper()
+        self._allowed = set(allowed_methods)  # type: Set[str]
```
Comment (webknjaz) [inline aiohttp/web_exceptions.py:297]: Should `allowed_methods` be uppercased as well?

File: aiohttp/web_exceptions.py (hunk 6 above)
Comment (webknjaz) [reply aiohttp/web_exceptions.py]: Dunno, but it's ugly like this now:(

File: aiohttp/web_exceptions.py (hunk 5 above)
Comment (webknjaz) [reply aiohttp/web_exceptions.py]: Isn't it faster in case of exception handling branch invocation?

File: aiohttp/web_exceptions.py (hunk 6 above)
Comment (asvetlov) [reply aiohttp/web_exceptions.py]: Black reformatting fails on python 3.5.
Will apply it when ready

File: aiohttp/web_exceptions.py (hunk 7 above)
Comment (asvetlov) [reply aiohttp/web_exceptions.py:297]: Good question.
HTTP verbs are case sensitive by RFC.
Maybe even `method.upper()` is redundant.
aiohttp dispatcher never raises the lowercased method.

File: aiohttp/web_exceptions.py (hunk 5 above)
Comment (asvetlov) [reply aiohttp/web_exceptions.py]: Are you talking about `lru_cache` decorator from `functools`? This decorator is not for free.
Anyway, I'll switch to direct `HTTPStatus` usage.

File: aiohttp/web_exceptions.py [hunk 8]
```diff
@@ -410,3 +468,19 @@ class HTTPNotExtended(HTTPServerError):
 
 class HTTPNetworkAuthenticationRequired(HTTPServerError):
     status_code = 511
+
+
+def _initialize_default_reason() -> None:
+    for obj in globals().values():
```
Comment (webknjaz) [inline aiohttp/web_exceptions.py:473]: How about using
```suggestion
    for obj in HTTPException.__subclasses__():
```
which gets immediate children of the class.

Comment (asvetlov) [reply aiohttp/web_exceptions.py:473]: Interesting, but we need a recursive iteration to collect all classes

Comment (webknjaz) [reply aiohttp/web_exceptions.py:473]: Like here? https://stackoverflow.com/a/3862957/595220
Such helper looks quite clean. If we were on Python 3.6 I'd suggest `__init_subclass__()` for such tracking instead...

File: aiohttp/web_exceptions.py [hunk 9]
```diff
@@ -410,3 +468,19 @@ class HTTPNotExtended(HTTPServerError):
...
+        if isinstance(obj, type) and issubclass(obj, HTTPException):
```
[the remark this answers is not in the document]
Comment (asvetlov) [reply aiohttp/web_exceptions.py:474]: The first argument of `issubclass` should be a type:
```
>>> issubclass(1, int)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: issubclass() arg 1 must be a class
```
`globals()` have many values, not all of them are classes

**Comment by asvetlov** [comment]
@webknjaz thanks for the review.
Now I'm working on the documentation update.
Will address notes a little later.

File: aiohttp/web_exceptions.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (webknjaz) [reply aiohttp/web_exceptions.py:474]: Oh... How about `__subclasses__()` recursive traversal?

File: aiohttp/web_exceptions.py (hunk 8 above)
Comment (asvetlov) [reply aiohttp/web_exceptions.py:473]: Let's keep the code as is

**Comment by asvetlov** [comment]
Everything is done from my perspective

**Pull request merged**